challenge_demo/test/grammar_test.py

```diff
@@ -0,0 +1,130 @@
+from __future__ import print_function
+import collections
+import unittest
+
+from grammar_parser.cfgparser import CFGParser
+from robocup_knowledge import load_knowledge
+from robot_skills.mockbot import Mockbot
+from action_server.actions.action import ConfigurationResult
+from action_server.task_manager import TaskManager
+
+TMTestResult = collections.namedtuple("TMTestResult", ["recipe", "config_result"])
+
+
+class GrammarTest(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        cls.knowledge = load_knowledge('challenge_demo')
+        cls.parser = CFGParser.fromstring(cls.knowledge.grammar)
+        cls.task_manager = TaskManager(robot=Mockbot())
+
+        # ToDo: Should we make everything recursive???
+
+        # Rule corresponding to the grammar target
+        root_rule = cls.parser.rules[cls.knowledge.grammar_target]
+
+        # In principle, a list of actions can be provided to the actions server
+        # Therefore, we need to take to steps to list all actions
+        # I.e., the following actions_list_target refers to the grammar target
+        # containing a list of actions
+        actions_list_target = ""
+        for option in root_rule.options:
+            if "actions" not in option.lsemantic:
+                continue
+            actions_list_target = option.conjuncts[0].name  # This is too hardcoded
+            break
+
+        assert actions_list_target, "Cannot derive actions from grammar"
+        actions_list_rule = cls.parser.rules[actions_list_target]
+        assert len(actions_list_rule.options) == 1, \
+            "Don't know how to continue if the actions list rule has multiple options"
+        assert len(actions_list_rule.options[0].conjuncts) == 1, \
+            "Don't know how to continue if the actions options list rule has multiple conjunctions"
+
+        # Now, we need to get all separate options
+        actions_target = actions_list_rule.options[0].conjuncts[0].name
+
+        cls.actions_rule = cls.parser.rules[actions_target]
+
+    def test_grammar(self):
+
+        test_results = {}
+        for option in self.actions_rule.options:
```

Comment:
it is assumed that self.actions_rule.options is none empty. Should that be checked to be able to give a warning?

Reply:
It's possible to iterate over an empty list. E.g.,
```python
for item in []:
    print(item)
```
doesn't crash. However, this would mean that the code doesn't test anything...

Reply:
I added a check in eb78a9547f16cc368838786a8024ebe356a18b89 that tests if this test actually does something.